File: Programmers/sort1.py
def quick_sort(array, start, end):
    if start > end:
        return
    pivot = start
    left = start + 1
    right = end

    while left <= right:
        while left <= end and array[pivot] >= array[left]:
            left += 1
        while right > start and array[pivot] <= array[right]:
            right -= 1
        if left > right:
            array[pivot], array[right] = array[right], array[pivot]
        else:
            array[left], array[right] = array[right], array[left]

    quick_sort(array, start, right - 1)
    quick_sort(array, right + 1, end)


def solution(array, commands):
    answer = []

    for command in commands:
        sub_array = array[command[0] - 1:command[1]]
        quick_sort(sub_array, 0, len(sub_array) - 1)
        answer.append(sub_array[command[2] - 1])
    return answer

array = [1, 5, 2, 6, 3, 7, 4]
commands = [[2, 5, 3], [4, 4, 1], [1, 7, 3]]
# answer = [5, 6, 3]
print(solution(array, commands))

File: Programmers/sort1.js
function mergeSort(array) {
  const n = array.length;

  if (n <= 1) return;

  const mid = parseInt(n / 2);
  const g1 = array.slice(0, mid);
  const g2 = array.slice(mid, n);
  mergeSort(g1);
  mergeSort(g2);

  let i1 = 0;
  let i2 = 0;
  let ia = 0;
  while (i1 < g1.length && i2 < g2.length) {
    if (g1[i1] < g2[i2]) {
      array[ia] = g1[i1];
      i1++;
    } else {
      array[ia] = g2[i2];
      i2++;
    }
    ia++;
  }

  for (let i = i1; i < g1.length; i++) {
    array[ia++] = g1[i];
  }
  for (let i = i2; i < g2.length; i++) {
    array[ia++] = g2[i];
  }
}

function solution(array, commands) {
  var answer = [];

  for (let command of commands) {
    const sliceArray = array.slice(command[0] - 1, command[1]);
    mergeSort(sliceArray, 0, command[1] - command[0] + 1);
    answer.push(sliceArray[command[2] - 1]);
  }

  return answer;
}

let array = [1, 5, 2, 6, 3, 7, 4];
let commands = [
  [2, 5, 3],
  [4, 4, 1],
  [1, 7, 3],
];
// answer = [5, 6, 3]
console.log(solution(array, commands));
